kdump

dperpeet opened the pull request:
- [x] local filesystem dump target
- [x] test settings before writing config
- [x] get system service status of kdump correctly
- [x] detect if kdump is enabled at boot time, show hint how to enable if not
- [x] get reserved memory
- [x] toggle compression
- [x] integration tests
- [x] [Video on Youtube](https://youtu.be/VXhuqPR2K5c)

andreasn commented:
Looks like a good start. A couple of css issues that I need to take a closer look at.

I tried to change to non-local storage, but got this:
![screenshot from 2017-01-10 15-49-52](https://cloud.githubusercontent.com/assets/209895/21810516/72a6a7d8-d74c-11e6-8b65-99b64ac1c192.png)

andreasn commented:
These tooltips looks different from each other.

![screenshot from 2017-01-10 15-51-14](https://cloud.githubusercontent.com/assets/209895/21810632/dcc29046-d74c-11e6-9d05-ce7664d5bcfc.png)

andreasn commented:
Does this mean I have no reserved memory?

![screenshot from 2017-01-10 15-54-15](https://cloud.githubusercontent.com/assets/209895/21810685/15db80a4-d74d-11e6-92ae-09140733ea4f.png)

dperpeet commented:
> Does this mean I have no reserved memory?

Added a hint what to do in this case.



> These tooltips looks different from each other.

Yes, there are still issues with the popover. The css needs to be changed.



> I tried to change to non-local storage, but got this:

It is currently scoped down to support only local storage. The menu has an "other" entry because we don't want to overwrite those when writing the config.

andreasn commented:
it seems the text being off between the two rows is due to label adding an extra 5px as margin-bottom

andreasn commented:
I think we worked around this in the network dialogs and elsewhere.

andreasn commented:
`.form-table-ct td:first-child label {
	margin-bottom: 0;
}`

Will fix the uneven rows

andreasn commented:
The misalignment in the location dialog seems to be due to <label> being around the <input id="kdump-settings-local-directory">

andreasn commented:
So with regards to the Other option in the dropdown. I'm still sceptical about the mere existence of this one, since if you're setting the location in kdump.conf, why not set the compression there as well?
But I'm interested in hearing what others think as well. @stefwalter @garrett

![screenshot from 2017-01-12 18-08-24](https://cloud.githubusercontent.com/assets/209895/21899862/22434880-d8f2-11e6-83f4-5bf41ffa66d9.png)

If we keep it, I think it makes sense to put the text directly in the dropdown, so:
Location [ Use the setting in /etc/kdump.conf. \/ ]

andreasn commented:
I managed to get the checkbox under control with this html
`<div class="checkbox">
    <label>
      <input type="checkbox">Compress crash dumps to save space
    </label>
  </div>`

dperpeet commented:
> I managed to get the checkbox under control with this html
> <div class="checkbox"><label><input type="checkbox">Compress crash dumps to save space</label></div>

This works, but the text with the checkbox is lighter now. I also added a margin to align the checkbox itself a bit better vertically.



> .form-table-ct td:first-child label { margin-bottom: 0; }
> 
> Will fix the uneven rows

thanks, used



> The misalignment in the location dialog seems to be due to <label> being around the <input id="kdump-settings-local-directory">
> 
> placeholder="/var/crash" then needs to be value="/var/crash" to not look weird

changed

andreasn commented:
There is a big white empty space on the right side of the dialog, so I think it would be a good idea to use the class cockpit-modal-md from system.css (is that global btw?) to make the dialog 400px wide.
![screenshot from 2017-01-13 14-10-23](https://cloud.githubusercontent.com/assets/209895/21931138/0b40e8a6-d99a-11e6-85fe-217dbfc462a0.png)

dperpeet commented:
OnOff switch related changes moved to #5752

dperpeet commented:
> If we keep it, I think it makes sense to put the text directly in the dropdown, so:
> Location [ Use the setting in /etc/kdump.conf. / ]

This is what it looks like now:
![screenshot from 2017-01-17 16-18-00](https://cloud.githubusercontent.com/assets/8711649/22026271/974cd6d2-dcd0-11e6-9221-dfb91ad4d162.png)

andreasn commented:
> This is what it looks like now

Yeah, I think that makes sense.

dperpeet commented:
bot: Image refresh for fedora-25

cockpituous commented:
Image creation for fedora-25 in process on cockpit-9.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-fedora-25-1d5e19bf9344f9fcff9b327ce5fc8115f96e1618/

dperpeet commented:
bot: Image refresh for debian-8

cockpituous commented:
Image creation for debian-8 in process on cockpit-10.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-debian-8-1d5e19bf9344f9fcff9b327ce5fc8115f96e1618/

dperpeet commented:
bot: Image refresh for rhel-7

cockpituous commented:
Image creation done: https://github.com/cockpituous/cockpit/commits/refresh-debian-8-1d5e19bf9344f9fcff9b327ce5fc8115f96e1618

cockpituous commented:
Image creation aborted

dperpeet commented:
bot: Image refresh for fedora-25

cockpituous commented:
Image creation for fedora-25 in process on cockpit-8.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-fedora-25-1d5e19bf9344f9fcff9b327ce5fc8115f96e1618/

cockpituous commented:
Image creation aborted

dperpeet commented:
bot: Image refresh for rhel-7

dperpeet commented:
integration tests should currently run on debian-8, rhel-atomic and fedora-atomic

dperpeet commented:
bot: Image refresh for fedora-25

stefwalter commented:
This needs a video on Youtube before it gets merged.

stefwalter commented:
```
# testBasic (check_kdump.TestKdump)
#
Failed to execute operation: No such file or directory
Resource Error: Error downloading http://10.111.117.14:9090/cockpit/$bb0d378490b6e70ba8642cb3c0154cdb47c33752/kdump/index.html#/ - server replied: Not Found http://10.111.117.14:9090/cockpit/$bb0d378490b6e70ba8642cb3c0154cdb47c33752/kdump/index.html#/
Wrote TestKdump-testBasic-FAIL.png
Journal extracted to TestKdump-testBasic-10.111.117.14-FAIL.log
not ok 62 testBasic (check_kdump.TestKdump) duration: 250s
Traceback (most recent call last):
  File "./verify/check-kdump", line 67, in testBasic
    self.login_and_go("/kdump")
  File "/build/cockpit/test/common/testlib.py", line 538, in login_and_go
    self.browser.login_and_go(path, user=user, host=host, authorized=authorized)
  File "/build/cockpit/test/common/testlib.py", line 368, in login_and_go
    self.enter_page(path.split("#")[0], host=host)
  File "/build/cockpit/test/common/testlib.py", line 317, in enter_page
    self.wait_present("iframe.container-frame[name='%s'][data-loaded]" % frame)
  File "/build/cockpit/test/common/testlib.py", line 215, in wait_present
    return self.wait_js_func('ph_is_present', selector)
  File "/build/cockpit/test/common/testlib.py", line 209, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 719, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 742, in _invoke
    raise Error(res['error'])
Error: timeout

# 3 TESTS FAILED [304s on cockpit-9]
```

stefwalter commented on pkg/kdump/config-client.es6:
```diff
@@ -0,0 +1,165 @@
+export class ConfigFile {
+    constructor(filename, superuser = false) {
+        this.filename = filename;
+        this._rawContent = undefined;
```
Is this really how private variables are done in ES6? There's no better way than in Javascript? If not, why not just use private scopes the same way?

dperpeet replied:
There is currently no concept of a `private` variable. Even with [workarounds](http://www.2ality.com/2016/01/private-data-classes.html) the data is still technically accessible.

stefwalter commented on pkg/kdump/config-client.es6:
```diff
@@ -0,0 +1,165 @@
+        if (this._dataAvailable)
+            this._dataAvailable.resolve();
+
+        var self = this;
```
Is this really necessary in ES6? Shouldn't ```this``` be more reliable and stably accessible?

dperpeet replied:
Even when adding the other function for a watch?

dperpeet replied:
I removed all the `self` use in the `.es6` files, but I had to add a ` /* jshint esversion:6 */` to get it to work with jshint.

stefwalter commented on pkg/kdump/config-client.es6:
```diff
@@ -0,0 +1,165 @@
+        // make sure we copy the original keys so we overwrite the correct lines when saving
+        self._originalSettings = { };
+        Object.keys(self.settings).forEach(function(key) {
+            self._originalSettings[key] = cockpit.extend({}, self.settings[key]);
```
ES6 hash Object.assign() which can be used instead of cockpit.extend(). It doesn't support Internet Explorer natively, but  since we're compiling via Babel won't an alternate implementation be brought in?

Worth checking.

dperpeet replied:
Yes, that seems to have a [polyfill](https://babeljs.io/docs/plugins/transform-object-assign/) when transforming.

dperpeet replied:
I had to change this back to using `cockpit.extend` since the unit test for the config client fails when using `Object.assign`

stefwalter commented on pkg/kdump/config-client.es6:
```diff
@@ -0,0 +1,165 @@
+
+    /* generate the config file from raw text and settings
+     */
+    _generateConfig(settings) {
```
Again, ditto for private functions. If there's no better way for doing these in ES6 ... then there's no real value to ES6 classes ... and I would suggest just sticking with the usual javascript definition of Objects + methods + properties + private scope.

dperpeet replied:
I prefer the ES6 class syntax, including `constructor`, not having to separate functions with a comma and the option of using getters and setters.

stefwalter commented on pkg/kdump/kdump-client.es6:
```diff
@@ -0,0 +1,183 @@
+            if (filename.slice(-1) != "/")
+                filename += "/";
+            filename += "cockpit.write.tmp";
+            cockpit.spawn(["/bin/sh", "-c", "touch " + filename + " && rm " + filename], { superuser: "try" } )
```
Which directory does this happen in? Should it just happen in the home directory? Likely /var/crash or /tmp or $TEMP is better, no?

stefwalter replied:
Worth using cockpit.script()?

dperpeet replied:
`filename` is the full path + filename, see [here](https://github.com/cockpit-project/cockpit/pull/5587/files#diff-43e57466c96b7104a2074b6470253920R103)
would it help to rename this to `filepath`?

dperpeet replied:
this has to make sure that the chosen directory exists and is writable

dperpeet replied:
yes, I think that will help

stefwalter replied:
Aha makes sense.

stefwalter commented on pkg/kdump/kdump-client.es6:
```diff
@@ -0,0 +1,183 @@
+    }
+    crashKernel() {
+        // crash the system kernel
+        return cockpit.spawn(["/bin/sh", "-c", "echo 1 > /proc/sys/kernel/sysrq && echo c > /proc/sysrq-trigger"], { superuser: "require" });
```
Lets split this out into multiple lines so it's more readable.

stefwalter replied:
Worth using cockpit.script()?

stefwalter replied:
At a very minimum ... This need to sync disks the same way you do it in the tests.

dperpeet replied:
changed

stefwalter commented on pkg/kdump/kdump-client.es6:
```diff
@@ -0,0 +1,183 @@
+
+        var dfd = cockpit.defer();
+        var filename = path;
+        if (!target || target.target != "local")
```
If we use braces in the else, I would suggest using them here too.

dperpeet replied:
changed

stefwalter commented on pkg/kdump/kdump-view.jsx:
```diff
@@ -0,0 +1,509 @@
+        this.props.onApplySettings(this.state.dialogSettings)
+            .done(function() {
+                dfd.resolve();
+
```
Extra line?

dperpeet replied:
changed

stefwalter commented on tools/cockpit.spec:
```diff
@@ -456,6 +459,17 @@ test -f %{_bindir}/firewall-cmd && firewall-cmd --reload --quiet || true
+Summary: Cockpit user interface for kernel crash dumping
+Requires: %{name}-bridge >= %{required_base}
+Requires: %{name}-shell >= %{required_base}
+BuildArch: noarch
```
This requires ```kexec-tools```. In addition cockpit-system on RHEL requires ```kexec-tools``` due to above.

```
 [stef@falcon cockpit]$ rpm -qf /usr/sbin/makedumpfile
kexec-tools-2.0.12-7.fc24.1.x86_64
```

dperpeet replied:
changed, also cockpit-system will provide cockpit-kdump on RHEL

stefwalter commented on tools/cockpit.spec:
```diff
@@ -456,6 +459,17 @@ test -f %{_bindir}/firewall-cmd && firewall-cmd --reload --quiet || true
+The Cockpit component for configuring kernel crash dumping.
+
+%files kdump -f kdump.list
+
```
The debian files need to build and install cockpit-kdump ... if that indeed is to be delivered there. It seems to be the case since you built a new Debian image.

dperpeet replied:
moving debian testing to follow-up

dperpeet commented:
bot: Image refresh for fedora-25

cockpituous commented:
Image creation for fedora-25 in process on cockpit-11.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-fedora-25-f5c0bb2d8dc84bc068b2f77ed84c0c4384e0dc8d/

dperpeet commented:
all changes implemented, but waiting for new images

cockpituous commented:
Image creation done: https://github.com/cockpituous/cockpit/commits/refresh-fedora-25-f5c0bb2d8dc84bc068b2f77ed84c0c4384e0dc8d

dperpeet commented:
`check-kdump` passed on rhel, rhel-atomic and fedora-25
I'll retrigger to make sure it's reliable.

dperpeet commented:
fedora-atomic is failing due to missing `kexec-tools`

dperpeet commented:
the new rhel-7 and fedora-25 images have an atomic scan failure, @larskarlitski is looking into that

larskarlitski commented:
> the new rhel-7 and fedora-25 images have an atomic scan failure, @larskarlitski is looking into that

This is due to a recently-introduced bug in atomic-scan that I proposed a [fix](https://github.com/projectatomic/atomic/pull/817) for when we noticed it in fedora-testing. Unfortunately, upstream developers didn't merge the fix, yet.

dperpeet commented:
> fedora-atomic is failing due to missing kexec-tools

neither installing nor testing cockpit-kdump on fedora-atomic

dperpeet commented:
rebased due to new fedora-24 image on master

stefwalter commented:
I think this commit needs:

```
diff --git a/test/verify/naughty-fedora-25/5750-atomic-storage-reset-regressed b/test/verify/naughty-fedora-25/5750-atomic-storage-reset-regressed
new file mode 100644
index 0000000..1e9ead9
--- /dev/null
+++ b/test/verify/naughty-fedora-25/5750-atomic-storage-reset-regressed
@@ -0,0 +1,5 @@
+Traceback (most recent call last):
+  File "./verify/check-docker-storage", line *, in testDevmapperVGroup
+    self.testDevmapper("dockah")
+  File "./verify/check-docker-storage", line *, in testDevmapper
+    b.wait_not_present(".modal-dialog:contains(Add Additional Storage)")
```

stefwalter commented:
Added a commit for the Fedora 25 issue.

dperpeet commented:
> Added a commit for the Fedora 25 issue.

I had to change this a bit

larskarlitski commented on pkg/kdump/kdump.js:
```diff
@@ -0,0 +1,116 @@
+                else
+                    dataStore.kdumpMemory = cockpit.format_bytes(value*1024*1024);
+            } else {
+                dataStore.kdumpMemory = content.trim();
```
Which values can be `kexec_crash_size`? I thought that there's always a value in bytes? Do we need the guessing the size then?

Note that `parseInt("256M", 10)` would also return `256`.

dperpeet replied:
See my comment above, the size can be either in bytes or megabytes, but useful values are <1G generally. The output is a raw number without units, but from the size of the number we can guess the unit.

larskarlitski commented on pkg/kdump/config-client.es6:
```diff
@@ -0,0 +1,163 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2016 Red Hat, Inc.
```
Happy new year!

dperpeet replied:
these files were created last year, so technically they're still accurate...

dperpeet commented:
> I noticed that the "Crash dump location" spinner is running forever when kexec-tools is not installed. Clicking the "Loading" link gives an Oops in that case.
That shouldn't happen, really, but I can look into a timeout. The cockpit-package depends on `kexec-tools`.
 
> Please add kexec-tools to the Vagrantfile as well, on line 45.
ok

> I cannot toggle the "Compress crash dumps to save space". How can go about debugging this?
Good catch, regression.

larskarlitski commented:
> That shouldn't happen, really, but I can look into a timeout. The cockpit-package depends on kexec-tools.

Ah, in that case there's no need to add it to the Vagrantfile.

dperpeet commented:
> I noticed that the "Crash dump location" spinner is running forever when kexec-tools is not installed. Clicking the "Loading" link gives an Oops in that case.

I cleaned up some related use cases and there was an actual bug in the way the config was being parsed. New behavior when kexec-tools isn't installed:

![screenshot from 2017-01-24 17-59-17](https://cloud.githubusercontent.com/assets/8711649/22257788/5f5ddd72-e25f-11e6-8367-cef9e37fac60.png)

And when the service is disabled, crashing the kernel for testing is now also disabled:
![screenshot from 2017-01-24 17-59-22](https://cloud.githubusercontent.com/assets/8711649/22257806/6a09916c-e25f-11e6-99b0-ee6e1f3d4cd3.png)

larskarlitski commented:
Thanks for the quick fixes @dperpeet!

martinpitt commented:
The fedora-25 failure looks relevant?
```
# testBasic (check_kdump.TestKdump)
#
Created symlink /etc/systemd/system/multi-user.target.wants/kdump.service  /usr/lib/systemd/system/kdump.service.
Job for kdump.service failed because the control process exited with error code.
See "systemctl status kdump.service" and "journalctl -xe" for details.
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-4.9.3-200.fc25.x86_64
Found initrd image: /boot/initramfs-4.9.3-200.fc25.x86_64.img
Found linux image: /boot/vmlinuz-4.8.6-300.fc25.x86_64
Found initrd image: /boot/initramfs-4.8.6-300.fc25.x86_64.img
Found linux image: /boot/vmlinuz-0-rescue-1a39617d347a459a988c278bfc248005
Found initrd image: /boot/initramfs-0-rescue-1a39617d347a459a988c278bfc248005.img
done
> Unable to apply settings: Directory /var/crash2 isn't writable or doesn't exist.
Page error: button.btn-default is disabled or somehow not clickable
phantomjs://code/phantom-lib.js 127 ph_click
 1 
 1 
Wrote TestKdump-testBasic-FAIL.png
Journal extracted to TestKdump-testBasic-10.111.122.112-FAIL.log
not ok 123 testBasic (check_kdump.TestKdump) duration: 112s
Traceback (most recent call last):
  File "./verify/check-kdump", line 110, in testBasic
    self.browser.click("button.btn-default")
  File "/build/cockpit/test/common/testlib.py", line 154, in click
    self.call_js_func('ph_click', selector, force)
  File "/build/cockpit/test/common/testlib.py", line 139, in call_js_func
    return self.phantom.eval("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 719, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 742, in _invoke
    raise Error(res['error'])
Error: button.btn-default is disabled or somehow not clickable
```

dperpeet commented:
> The fedora-25 failure looks relevant?

yes, it looks like a test race because of the newly disabled button I put in...

dperpeet commented:
Added a wait in the test to make sure the service is running properly before hitting the test button that crashes the system.

martinpitt commented:
The two rhel-7 test failures are unrelated, everything else is green now (nice!), and the new waiting looks good. Thanks!
